src/bases.jl

```diff
@@ -518,6 +506,94 @@ function _get_vectors(B::CachedBasis{𝔽,<:AbstractBasis,<:DiagonalizingBasisDa
     return B.data.vectors
 end
 
+
+@doc raw"""
+    gram_schmidt(M::Manifold, p, B::AbstractBasis)
+    gram_schmidt(M::Manifold, p, V::AbstractVector, field = number_system(M))
+
+Compute an ONB in the tangent space at `p` on the [`Manifold`](@ref} `M` from either an
+[`AbstractBasis`](@ref) basis ´B´ or a set of (at most) [`manifold_dimension`](@ref)`(M)`
+many vectors.
+The `number_system` is taken from the basis or manifold, but can also be set to indicate
+a different basis (i.e. a real basis on a complex manifold) for a set of vectors.
+
+The method always returns a basis, i.e. linearly dependent vectors are removed.
+
+# Keyword arguments
+
+* `warn_linearly_dependent` (`false`) – warn if the basis vectors are not linearly
+  independent
+* `return_incomplete_set` (`false`) – throw an error if the resulting set of vectors is not
+  a basis but contains less vectors
+
+further keyword arguments can be passed to set the accuracy of the independence test.
+
+# Return value
+
+When a set of vectors is orthonormalized a set of vectors is returned.
+When an [`AbstractBasis`](@ref) is orthonormalized, a [`CachedBasis`](@ref) is returned.
+"""
+function gram_schmidt(
+    M::Manifold,
+    p,
+    B::AbstractBasis{𝔽};
+    warn_linearly_dependent = false,
+    return_incomplete_set = false,
+    kwargs...,
+) where {𝔽}
+    V = gram_schmidt(
+        M,
+        p,
+        get_vectors(M, p, B),
+        number_system(B);
+        warn_linearly_dependent = warn_linearly_dependent,
+        return_incomplete_set = return_incomplete_set,
+        kwargs...,
+    )
+    return CachedBasis(GramSchmidtOrthonormalBasis(𝔽), V)
+end
+function gram_schmidt(
+    M::Manifold,
+    p,
+    V::AbstractVector,
+    field = number_system(M);
+    warn_linearly_dependent = false,
+    return_incomplete_set = false,
+    kwargs...,
+)
+    N = length(V)
+    Ξ = empty(V)
+    dim = manifold_dimension(M)
+    N < dim && @warn "Input only has $(N) vectors, but manifold dimension is $(dim)."
+    @inbounds for n in 1:N
+        Ξₙ = V[n]
+        for k in 1:length(Ξ)
+            Ξₙ .-= real(inner(M, p, Ξ[k], Ξₙ)) .* Ξ[k]
+        end
+        nrmΞₙ = norm(M, p, Ξₙ)
+        if nrmΞₙ == 0
+            warn_linearly_dependent && @warn "Input vector $(n) has length 0."
+            @goto skip
+        end
+        Ξₙ ./= nrmΞₙ
+        for k in 1:length(Ξ)
+            if !isapprox(real(inner(M, p, Ξ[k], Ξₙ)), 0; kwargs...)
+                warn_linearly_dependent &&
+                    @warn "Input vector $(n) is not linearly independent of output basis vector $(k)."
```

Comment:
@mateuszbaran I have a question on this error message. If the nth vector is in the span of the previous vectors, it should be zero after lines 570-572, so the 574 should be active.

So when actually would this method kick in? I haven't completely figured this out but the error message seems to vague, since we also do not compare the actual input vector but its “rework” (remainder after 571&normalized).

Reply:
IIRC this check is done like that to deal in some way with numerical error. I don't know if this is the best way to do that, likely not, but I think this makes sense. 574 only checks for exactly zero while 580 checks for near-zero inner product with one of the previous basis vectors. The check in 574 is I think just to make sure that division in 578 doesn't explode.

The error message indeed doesn't explain how the linear independence is tested, this could be improved.

Reply:
That sounds reasonable. Yes then mainly the error message should be improved. I hope my doc string covers all special cases you had in mind.

Reply:
Yes, the docstring looks more or less complete. I'll make a more detailed review later.